pkg/webhooks/validation/controller/controller.go

```diff
@@ -0,0 +1,530 @@
+	webhookInformer.AddEventHandler(makeHandler(c.queue, configGVK, o.WebhookConfigName))
+
+	endpointInformer := c.sharedInformers.Core().V1().Endpoints().Informer()
+	endpointInformer.AddEventHandler(makeHandler(c.queue, endpointGVK, o.ServiceName))
+
+	deploymentInformer := c.sharedInformers.Apps().V1().Deployments().Informer()
+	deploymentInformer.AddEventHandler(makeHandler(c.queue, deploymentGVK, o.GalleyDeploymentName))
+
+	return c, nil
+}
+
+func (c *Controller) Start(stop <-chan struct{}) {
+	go c.startFileWatcher(stop)
+	go c.sharedInformers.Start(stop)
+
+	for _, ready := range c.sharedInformers.WaitForCacheSync(stop) {
+		if !ready {
+			return
+		}
+	}
+
+	req := &reconcileRequest{"initial request to kickstart reconciliation"}
+	c.queue.Add(req)
+
+	go c.runWorker()
+}
+
+func (c *Controller) startFileWatcher(stop <-chan struct{}) {
+	for {
+		select {
+		case ev := <-c.caFileWatcher.Events(c.o.WebhookConfigPath):
+			req := &reconcileRequest{fmt.Sprintf("Webhook configuration changed: %v", ev)}
+			c.queue.Add(req)
+		case ev := <-c.caFileWatcher.Events(c.o.CAPath):
+			req := &reconcileRequest{fmt.Sprintf("CA file changed: %v", ev)}
+			c.queue.Add(req)
+		case err := <-c.caFileWatcher.Errors(c.o.WebhookConfigPath):
+			scope.Warnf("error watching local webhook configuration: %v", err)
+		case err := <-c.caFileWatcher.Errors(c.o.CAPath):
+			scope.Warnf("error watching local CA bundle: %v", err)
+		case <-stop:
+			return
+		}
+	}
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() (cont bool) {
+	obj, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	req, ok := obj.(*reconcileRequest)
+	if !ok {
+		// don't retry an invalid reconcileRequest item
+		c.queue.Forget(req)
+		return true
+	}
+
+	if err := c.reconcileRequest(req); err != nil {
+		c.queue.AddRateLimited(obj)
+		utilruntime.HandleError(err)
+	} else {
+		c.queue.Forget(obj)
+	}
+	return true
+}
+
+// reconcile the desired state with the kube-apiserver.
+func (c *Controller) reconcileRequest(req *reconcileRequest) error {
+	defer func() {
+		if c.reconcileDone != nil {
+			c.reconcileDone()
+		}
+	}()
+
+	scope.Infof("Reconcile: %v", req)
+
+	// don't create the webhook config before the endpoint is ready
+	if !c.endpointReadyOnce {
+		if ready, err := c.isEndpointReady(); !ready || err != nil {
+			scope.Infof("Endpoint not ready: ready=%v err=%v", ready, err)
+			return err
+		}
+		c.endpointReadyOnce = true
+	}
+
+	// don't update the webhook config if its already managed by an existing galley deployment.
+	if c.o.GalleyDeploymentName != "" {
+		if running, err := c.isGalleyDeploymentRunning(); running || err != nil {
+			scope.Infof("Galley deployment detected: running=%v err=%v", running, err)
+			return err
+		}
+	}
+
+	// actively remove the webhook configuration if the controller is running but the webhook
+	if c.o.UnregisterValidationWebhook {
```

Comment:
When is this used? It seems to always be false even in pr 3/3

Reply:
It is hooked up in PR 2/3 but I haven't pushed those latest changes yet.

Reply:
This would be used when someone re-configures Istio with validation disabled. We need to run the controller (Enabled=true) but force the reconciliation logic to remove the config. We had previously used a config with no rules when validation was disabled, but some users found it confusing to see a validatingwebhookconfiguration exist after validation was disabled.

Reply:
I see. I think that makes sense. Somewhat related: https://github.com/istio/istio/issues/19873. Basically there is a difference between deleting the deployment and disabling validation, so I think we still need a way to remove it in operator as well? Or I guess longer term operator itself may be doing this

Reply:
I'd expect the operator or istioctl to handle webook config lifecycle if galley/pilot isn't.